lightning/src/ln/channelmanager.rs

```diff
@@ -71,6 +71,265 @@ use core::time::Duration;
+		let hmac_bytes = Hmac::from_engine(hmac).into_inner();
+
+		let mut iv_bytes = [0 as u8; IV_LEN];
+		iv_bytes.copy_from_slice(&hmac_bytes[..IV_LEN]);
+
+		Ok(construct_payment_secret(&iv_bytes, &metadata_bytes, &keys.metadata_key))
+	}
+
+	fn construct_metadata_bytes(min_value_msat: Option<u64>, payment_type: Method, invoice_expiry_delta_secs: u32, highest_seen_timestamp: u64) -> Result<[u8; METADATA_LEN], ()> {
+		if min_value_msat.is_some() && min_value_msat.unwrap() > MAX_VALUE_MSAT {
+			return Err(());
+		}
+
+		let mut min_amt_msat_bytes: [u8; AMT_MSAT_LEN] = match min_value_msat {
+			Some(amt) => amt.to_be_bytes(),
+			None => [0; AMT_MSAT_LEN],
+		};
+		min_amt_msat_bytes[0] |= (payment_type as u8) << METHOD_TYPE_OFFSET;
+
+		// We assume that highest_seen_timestamp is pretty close to the current time - it's updated when
+		// we receive a new block with the maximum time we've seen in a header. It should never be more
+		// than two hours in the future.  Thus, we add two hours here as a buffer to ensure we
+		// absolutely never fail a payment too early.
+		// Note that we assume that received blocks have reasonably up-to-date timestamps.
+		let expiry_bytes = (highest_seen_timestamp + invoice_expiry_delta_secs as u64 + 7200).to_be_bytes();
+
+		let mut metadata_bytes: [u8; METADATA_LEN] = [0; METADATA_LEN];
+		{
+			let (min_amt_msat_slice, expiry_slice) = metadata_bytes.split_at_mut(AMT_MSAT_LEN);
+			min_amt_msat_slice.copy_from_slice(&min_amt_msat_bytes);
+			expiry_slice.copy_from_slice(&expiry_bytes);
+		}
+
+		Ok(metadata_bytes)
+	}
+
+	fn construct_payment_secret(iv_bytes: &[u8; IV_LEN], metadata_bytes: &[u8], metadata_key: &[u8; METADATA_KEY_LEN]) -> PaymentSecret {
+		let mut payment_secret_bytes: [u8; 32] = [0; 32];
+		{
+			let (iv_slice, encrypted_metadata_slice) = payment_secret_bytes.split_at_mut(16);
+			iv_slice.copy_from_slice(iv_bytes);
+
+			let chacha_block = ChaCha20::get_single_block(metadata_key, iv_bytes);
+			for i in 0..METADATA_LEN {
+				encrypted_metadata_slice[i] = chacha_block[i] ^ metadata_bytes[i];
+			}
+		}
+		PaymentSecret(payment_secret_bytes)
+	}
+
+	/// Check that an inbound payment's `payment_data` field is sane.
+	///
+	/// LDK does not store any data for pending inbound payments. Instead, we construct our payment
+	/// secret (and, if supplied by LDK, our payment preimage) to include encrypted metadata about the
+	/// payment.
+	///
+	/// The metadata is constructed as:
+	///   payment method (3 bits) || payment amount (8 bytes - 3 bits) || expiry (8 bytes)
+	/// and encrypted using a key derived from [`KeysInterface::get_inbound_payment_key_material`].
+	///
+	/// Then on payment receipt, we verify in this method that the payment preimage and payment secret
+	/// match what was constructed.
+	///
+	/// [`create_inbound_payment`] and [`create_inbound_payment_for_hash`] are called by the user to
+	/// construct the payment secret and/or payment hash that this method is verifying.
```

Comment:
Could you reword this slightly to tie the fact that these are the payment methods represented by the `Method` enum?